// framework7-react/cjs/shared/use-tooltip.js
"use strict";

exports.__esModule = true;
exports.useTooltip = void 0;

var _react = require("react");

var _watchProp = require("./watch-prop");

var _f = require("./f7");

var useTooltip = function useTooltip(elRef, props) {
  var f7Tooltip = (0, _react.useRef)(null);
  var tooltip = props.tooltip,
      tooltipTrigger = props.tooltipTrigger;

  var onMount = function onMount() {
    if (!elRef.current) return;
    if (!tooltip) return;
    (0, _f.f7ready)(function () {
      f7Tooltip.current = _f.f7.tooltip.create({
        targetEl: elRef.current,
        text: tooltip,
        trigger: tooltipTrigger
      });
    });
  };

  var onDestroy = function onDestroy() {
    if (f7Tooltip.current && f7Tooltip.current.destroy) {
      f7Tooltip.current.destroy();
      f7Tooltip.current = null;
    }
  };

  (0, _react.useEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  (0, _watchProp.watchProp)(tooltip, function (value) {
    if (!value && f7Tooltip.current) {
      f7Tooltip.current.destroy();
      f7Tooltip.current = null;
      return;
    }

    if (value && !f7Tooltip.current && _f.f7) {
      f7Tooltip.current = _f.f7.tooltip.create({
        targetEl: elRef.current,
        text: value,
        trigger: tooltipTrigger
      });
      return;
    }

    if (!value || !f7Tooltip.current) return;
    f7Tooltip.current.setText(value);
  });
};

exports.useTooltip = useTooltip;
