// framework7-react/esm/components/list-index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, emit } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { f7ready, f7 } from '../shared/f7';
import { watchProp } from '../shared/watch-prop';

var ListIndex = /*#__PURE__*/forwardRef(function (props, ref) {
  var _f7ListIndex = useRef(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      _props$init = props.init,
      init = _props$init === void 0 ? true : _props$init,
      listEl = props.listEl,
      _props$indexes = props.indexes,
      indexes = _props$indexes === void 0 ? 'auto' : _props$indexes,
      _props$scrollList = props.scrollList,
      scrollList = _props$scrollList === void 0 ? true : _props$scrollList,
      _props$label = props.label,
      label = _props$label === void 0 ? false : _props$label,
      _props$iosItemHeight = props.iosItemHeight,
      iosItemHeight = _props$iosItemHeight === void 0 ? 14 : _props$iosItemHeight,
      _props$mdItemHeight = props.mdItemHeight,
      mdItemHeight = _props$mdItemHeight === void 0 ? 14 : _props$mdItemHeight,
      _props$auroraItemHeig = props.auroraItemHeight,
      auroraItemHeight = _props$auroraItemHeig === void 0 ? 14 : _props$auroraItemHeig;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);

  var update = function update() {
    if (!_f7ListIndex.current) return;

    _f7ListIndex.current.update();
  };

  var scrollListToIndex = function scrollListToIndex(indexContent) {
    if (!_f7ListIndex.current) return;

    _f7ListIndex.current.scrollListToIndex(indexContent);
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      f7ListIndex: function f7ListIndex() {
        return _f7ListIndex.current;
      },
      update: update,
      scrollListToIndex: scrollListToIndex
    };
  });
  watchProp(indexes, function (newValue) {
    if (!_f7ListIndex.current) return;
    _f7ListIndex.current.params.indexes = newValue;
    update();
  });

  var onMount = function onMount() {
    if (!init) return;
    f7ready(function () {
      _f7ListIndex.current = f7.listIndex.create({
        el: elRef.current,
        listEl: listEl,
        indexes: indexes,
        iosItemHeight: iosItemHeight,
        mdItemHeight: mdItemHeight,
        auroraItemHeight: auroraItemHeight,
        scrollList: scrollList,
        label: label,
        on: {
          select: function select(index, itemContent, itemIndex) {
            emit(props, 'listIndexSelect', itemContent, itemIndex);
          }
        }
      });
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7ListIndex.current && _f7ListIndex.current.destroy) {
      _f7ListIndex.current.destroy();
    }

    _f7ListIndex.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  var classes = classNames(className, 'list-index', colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
ListIndex.displayName = 'f7-list-index';
export default ListIndex;
