// framework7-react/esm/components/login-screen.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, emit } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { f7ready, f7 } from '../shared/f7';
import { watchProp } from '../shared/watch-prop';
import { modalStateClasses } from '../shared/modal-state-classes';


var LoginScreen = /*#__PURE__*/forwardRef(function (props, ref) {
  var _f7LoginScreen = useRef(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      opened = props.opened,
      animate = props.animate,
      containerEl = props.containerEl;
  var extraAttrs = getExtraAttrs(props);
  var isOpened = useRef(opened);
  var isClosing = useRef(false);
  var elRef = useRef(null);

  var onOpen = function onOpen(instance) {
    isOpened.current = true;
    isClosing.current = false;
    emit(props, 'loginScreenOpen', instance);
  };

  var onOpened = function onOpened(instance) {
    emit(props, 'loginScreenOpened', instance);
  };

  var onClose = function onClose(instance) {
    isOpened.current = false;
    isClosing.current = true;
    emit(props, 'loginScreenClose', instance);
  };

  var onClosed = function onClosed(instance) {
    isClosing.current = false;
    emit(props, 'loginScreenClosed', instance);
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      f7LoginScreen: function f7LoginScreen() {
        return _f7LoginScreen.current;
      }
    };
  }); // watch opened changes

  watchProp(opened, function (value) {
    if (!_f7LoginScreen.current) return;

    if (value) {
      _f7LoginScreen.current.open();
    } else {
      _f7LoginScreen.current.close();
    }
  });

  var modalEvents = function modalEvents(method) {
    if (!_f7LoginScreen.current) return;

    _f7LoginScreen.current[method]('open', onOpen);

    _f7LoginScreen.current[method]('opened', onOpened);

    _f7LoginScreen.current[method]('close', onClose);

    _f7LoginScreen.current[method]('closed', onClosed);
  };

  var onMount = function onMount() {
    if (!elRef.current) return;
    f7ready(function () {
      var loginScreenParams = {
        el: elRef.current
      };
      if ('animate' in props) loginScreenParams.animate = animate;
      if ('containerEl' in props) loginScreenParams.containerEl = containerEl;
      _f7LoginScreen.current = f7.loginScreen.create(loginScreenParams);
      modalEvents('on');

      if (opened) {
        _f7LoginScreen.current.open(false);
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7LoginScreen.current) {
      _f7LoginScreen.current.destroy();
    }

    _f7LoginScreen.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    modalEvents('on');
    return function () {
      modalEvents('off');
    };
  });
  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  var classes = classNames(className, 'login-screen', modalStateClasses({
    isOpened: isOpened,
    isClosing: isClosing
  }), colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
LoginScreen.displayName = 'f7-login-screen';
export default LoginScreen;
