// framework7-react/esm/components/menu-dropdown-item.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils';
import { colorClasses, routerAttrs, routerClasses, actionsAttrs, actionsClasses } from '../shared/mixins';
import { useRouteProps } from '../shared/use-route-props';

var MenuDropdownItem = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      link = props.link,
      href = props.href,
      target = props.target,
      text = props.text,
      divider = props.divider,
      menuClose = props.menuClose;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);

  var onClick = function onClick(e) {
    emit(props, 'click', e);
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current
    };
  });
  useRouteProps(elRef, props);
  var isLink = link || href || href === '';
  var Tag = isLink ? 'a' : 'div';
  var classes = classNames({
    'menu-dropdown-link': isLink && !divider,
    'menu-dropdown-item': !isLink && !divider,
    'menu-dropdown-divider': divider
  }, className, colorClasses(props), routerClasses(props), actionsClasses(props), {
    'menu-close': typeof menuClose === 'undefined'
  });
  var hrefComputed = href;
  if (typeof hrefComputed === 'undefined' && link) hrefComputed = '#';

  var attrs = _extends({
    href: hrefComputed,
    target: target
  }, routerAttrs(props), actionsAttrs(props));

  return /*#__PURE__*/React.createElement(Tag, _extends({
    className: classes,
    id: id,
    style: style,
    ref: elRef
  }, attrs, extraAttrs, {
    onClick: onClick
  }), text, children);
});
MenuDropdownItem.displayName = 'f7-menu-dropdown-item';
export default MenuDropdownItem;
