// framework7-react/esm/components/routable-modals.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle, useState } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { f7events, f7routers, f7 } from '../shared/f7';

var RoutableModals = /*#__PURE__*/forwardRef(function (props, ref) {
  var _useState = useState([]),
      modals = _useState[0],
      _setModals = _useState[1];

  var elRef = useRef(null);
  var routerData = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      el: elRef.current
    };
  });

  var onMount = function onMount() {
    routerData.current = {
      modals: modals,
      el: elRef.current,
      setModals: function setModals(newModals) {
        _setModals([].concat(newModals));
      }
    };
    f7routers.modals = routerData.current;
  };

  var onDestroy = function onDestroy() {
    if (!routerData.current) return;
    f7routers.modals = null;
    routerData.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  useIsomorphicLayoutEffect(function () {
    if (!routerData.current || !f7) return;
    f7events.emit('modalsRouterDidUpdate', routerData.current);
  });
  return /*#__PURE__*/React.createElement("div", {
    ref: elRef,
    className: "framework7-modals"
  }, modals.map(function (_ref) {
    var ModalComponent = _ref.component,
        modalId = _ref.id,
        modalProps = _ref.props;
    return /*#__PURE__*/React.createElement(ModalComponent, _extends({
      key: modalId
    }, modalProps));
  }));
});
RoutableModals.displayName = 'f7-routable-modals';
export default RoutableModals;
