// framework7-react/esm/components/text-editor.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, getSlots, noUndefinedProps, emit } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { f7ready, f7 } from '../shared/f7';
import { watchProp } from '../shared/watch-prop';


var TextEditor = /*#__PURE__*/forwardRef(function (props, ref) {
  var _f7TextEditor = useRef(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      mode = props.mode,
      value = props.value,
      buttons = props.buttons,
      customButtons = props.customButtons,
      dividers = props.dividers,
      imageUrlText = props.imageUrlText,
      linkUrlText = props.linkUrlText,
      placeholder = props.placeholder,
      clearFormattingOnPaste = props.clearFormattingOnPaste,
      _props$resizable = props.resizable,
      resizable = _props$resizable === void 0 ? false : _props$resizable;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);

  var onChange = function onChange(editor, editorValue) {
    emit(props, 'textEditorChange', editorValue);
  };

  var onInput = function onInput(editor, editorValue) {
    emit(props, 'textEditorInput', editorValue);
  };

  var onFocus = function onFocus() {
    emit(props, 'textEditorFocus');
  };

  var onBlur = function onBlur() {
    emit(props, 'textEditorBlur');
  };

  var onButtonClick = function onButtonClick(editor, button) {
    emit(props, 'textEditorButtonClick', button);
  };

  var onKeyboardOpen = function onKeyboardOpen() {
    emit(props, 'textEditorKeyboardOpen');
  };

  var onKeyboardClose = function onKeyboardClose() {
    emit(props, 'textEditorKeyboardClose');
  };

  var onPopoverOpen = function onPopoverOpen() {
    emit(props, 'textEditorPopoverOpen');
  };

  var onPopoverClose = function onPopoverClose() {
    emit(props, 'textEditorPopoverClose');
  };

  var onInsertLink = function onInsertLink(editor, url) {
    emit(props, 'textEditorInsertLink', url);
  };

  var onInsertImage = function onInsertImage(editor, url) {
    emit(props, 'textEditorInsertImage', url);
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      f7TextEditor: function f7TextEditor() {
        return _f7TextEditor.current;
      }
    };
  });
  watchProp(value, function (newValue) {
    if (_f7TextEditor.current) {
      _f7TextEditor.current.setValue(newValue);
    }
  });

  var onMount = function onMount() {
    var params = noUndefinedProps({
      el: elRef.current,
      mode: mode,
      value: value,
      buttons: buttons,
      customButtons: customButtons,
      dividers: dividers,
      imageUrlText: imageUrlText,
      linkUrlText: linkUrlText,
      placeholder: placeholder,
      clearFormattingOnPaste: clearFormattingOnPaste,
      on: {
        change: onChange,
        input: onInput,
        focus: onFocus,
        blur: onBlur,
        buttonClick: onButtonClick,
        keyboardOpen: onKeyboardOpen,
        keyboardClose: onKeyboardClose,
        popoverOpen: onPopoverOpen,
        popoverClose: onPopoverClose,
        insertLink: onInsertLink,
        insertImage: onInsertImage
      }
    });
    f7ready(function () {
      _f7TextEditor.current = f7.textEditor.create(params);
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7TextEditor.current && _f7TextEditor.current.destroy) {
      _f7TextEditor.current.destroy();
    }

    _f7TextEditor.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  var slots = getSlots(props);
  var classes = classNames(className, 'text-editor', resizable && 'text-editor-resizable', colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: elRef,
    id: id,
    style: style,
    className: classes
  }, extraAttrs), slots['root-start'], /*#__PURE__*/React.createElement("div", {
    className: "text-editor-content",
    contentEditable: true
  }, slots.default), slots['root-end'], slots.root);
});
TextEditor.displayName = 'f7-text-editor';
export default TextEditor;
