// framework7-react/cjs/components/list-item.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _useRouteProps = require("../shared/use-route-props");

var _useSmartSelect = require("../shared/use-smart-select");

var _useTooltip = require("../shared/use-tooltip");

var _watchProp = require("../shared/watch-prop");

var _f = require("../shared/f7");

var _listItemContent = _interopRequireDefault(require("./list-item-content"));

var _listContext = require("../shared/list-context");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }



/*
const ListItemContent = ({
  props,
  slots,
  inputElRef,
  onChange,
  onClick,
  isMediaComputed,
  isSortableComputed,
  isSortableOppositeComputed,
} = {}) => {
  const {
    radio,
    checkbox,
    value,
    name,
    readonly,
    disabled,
    checked,
    defaultChecked,
    required,
    media,
    header,
    footer,
    title,
    subtitle,
    text,
    after,
    badge,
    badgeColor,
    radioIcon,
    swipeout,
    sortable,
    accordionItem,
  } = props;

};
*/
var ListItem = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      title = props.title,
      link = props.link,
      target = props.target,
      tabLink = props.tabLink,
      tabLinkActive = props.tabLinkActive,
      selected = props.selected,
      mediaItem = props.mediaItem,
      mediaList = props.mediaList,
      divider = props.divider,
      groupTitle = props.groupTitle,
      swipeout = props.swipeout,
      swipeoutOpened = props.swipeoutOpened,
      sortable = props.sortable,
      sortableOpposite = props.sortableOpposite,
      accordionItem = props.accordionItem,
      accordionItemOpened = props.accordionItemOpened,
      smartSelect = props.smartSelect,
      smartSelectParams = props.smartSelectParams,
      noChevron = props.noChevron,
      chevronCenter = props.chevronCenter,
      checkbox = props.checkbox,
      radio = props.radio,
      disabled = props.disabled,
      virtualListIndex = props.virtualListIndex,
      href = props.href;
  var listContext = (0, _react.useContext)(_listContext.ListContext);

  var _ref = listContext || {},
      _ref$listIsMedia = _ref.listIsMedia,
      listIsMedia = _ref$listIsMedia === void 0 ? false : _ref$listIsMedia,
      _ref$listIsSortable = _ref.listIsSortable,
      listIsSortable = _ref$listIsSortable === void 0 ? false : _ref$listIsSortable,
      _ref$listIsSortableOp = _ref.listIsSortableOpposite,
      listIsSortableOpposite = _ref$listIsSortableOp === void 0 ? false : _ref$listIsSortableOp,
      _ref$listIsSimple = _ref.listIsSimple,
      listIsSimple = _ref$listIsSimple === void 0 ? false : _ref$listIsSimple;

  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  var linkElRef = (0, _react.useRef)(null);

  var _f7SmartSelect = (0, _react.useRef)(null);

  var onClick = function onClick(event) {
    if (event.target.tagName.toLowerCase() !== 'input') {
      (0, _utils.emit)(props, 'click', event);
    }
  };

  var onSwipeoutOverswipeEnter = function onSwipeoutOverswipeEnter(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutOverswipeEnter');
  };

  var onSwipeoutOverswipeExit = function onSwipeoutOverswipeExit(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutOverswipeExit');
  };

  var onSwipeoutDeleted = function onSwipeoutDeleted(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutDeleted');
  };

  var onSwipeoutDelete = function onSwipeoutDelete(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutDelete');
  };

  var onSwipeoutClose = function onSwipeoutClose(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutClose');
  };

  var onSwipeoutClosed = function onSwipeoutClosed(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutClosed');
  };

  var onSwipeoutOpen = function onSwipeoutOpen(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutOpen');
  };

  var onSwipeoutOpened = function onSwipeoutOpened(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeoutOpened');
  };

  var onSwipeout = function onSwipeout(el, progress) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'swipeout', progress);
  };

  var onAccBeforeClose = function onAccBeforeClose(el, prevent) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionBeforeClose', prevent);
  };

  var onAccClose = function onAccClose(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionClose');
  };

  var onAccClosed = function onAccClosed(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionClosed');
  };

  var onAccBeforeOpen = function onAccBeforeOpen(el, prevent) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionBeforeOpen', prevent);
  };

  var onAccOpen = function onAccOpen(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionOpen');
  };

  var onAccOpened = function onAccOpened(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'accordionOpened');
  };

  var onChange = function onChange(event) {
    (0, _utils.emit)(props, 'change', event);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7SmartSelect: function f7SmartSelect() {
        return _f7SmartSelect.current;
      }
    };
  });
  (0, _useTooltip.useTooltip)(elRef, props);
  (0, _useRouteProps.useRouteProps)(linkElRef, props);
  (0, _watchProp.watchProp)(swipeoutOpened, function (newValue) {
    if (!swipeout || !elRef.current || !_f.f7) return;

    if (newValue) {
      _f.f7.swipeout.open(elRef.current);
    } else {
      _f.f7.swipeout.close(elRef.current);
    }
  });

  var attachEvents = function attachEvents() {
    (0, _f.f7ready)(function () {
      if (swipeout) {
        _f.f7.on('swipeoutOpen', onSwipeoutOpen);

        _f.f7.on('swipeoutOpened', onSwipeoutOpened);

        _f.f7.on('swipeoutClose', onSwipeoutClose);

        _f.f7.on('swipeoutClosed', onSwipeoutClosed);

        _f.f7.on('swipeoutDelete', onSwipeoutDelete);

        _f.f7.on('swipeoutDeleted', onSwipeoutDeleted);

        _f.f7.on('swipeoutOverswipeEnter', onSwipeoutOverswipeEnter);

        _f.f7.on('swipeoutOverswipeExit', onSwipeoutOverswipeExit);

        _f.f7.on('swipeout', onSwipeout);
      }

      if (accordionItem) {
        _f.f7.on('accordionBeforeOpen', onAccBeforeOpen);

        _f.f7.on('accordionOpen', onAccOpen);

        _f.f7.on('accordionOpened', onAccOpened);

        _f.f7.on('accordionBeforeClose', onAccBeforeClose);

        _f.f7.on('accordionClose', onAccClose);

        _f.f7.on('accordionClosed', onAccClosed);
      }
    });
  };

  var detachEvents = function detachEvents() {
    if (!_f.f7) return;

    _f.f7.off('swipeoutOpen', onSwipeoutOpen);

    _f.f7.off('swipeoutOpened', onSwipeoutOpened);

    _f.f7.off('swipeoutClose', onSwipeoutClose);

    _f.f7.off('swipeoutClosed', onSwipeoutClosed);

    _f.f7.off('swipeoutDelete', onSwipeoutDelete);

    _f.f7.off('swipeoutDeleted', onSwipeoutDeleted);

    _f.f7.off('swipeoutOverswipeEnter', onSwipeoutOverswipeEnter);

    _f.f7.off('swipeoutOverswipeExit', onSwipeoutOverswipeExit);

    _f.f7.off('swipeout', onSwipeout);

    _f.f7.off('accordionBeforeOpen', onAccBeforeOpen);

    _f.f7.off('accordionOpen', onAccOpen);

    _f.f7.off('accordionOpened', onAccOpened);

    _f.f7.off('accordionBeforeClose', onAccBeforeClose);

    _f.f7.off('accordionClose', onAccClose);

    _f.f7.off('accordionClosed', onAccClosed);
  };

  (0, _useSmartSelect.useSmartSelect)(smartSelect, smartSelectParams, _f7SmartSelect, function () {
    return elRef.current.querySelector('a.smart-select');
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    (0, _f.f7ready)(function () {
      if (swipeout && swipeoutOpened) {
        _f.f7.swipeout.open(elRef.current);
      }
    });
  }, []);
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    attachEvents();
    return detachEvents;
  });
  var slots = (0, _utils.getSlots)(props);
  var linkEl;
  var itemContentEl;
  var isMediaComputed = mediaItem || mediaList || listIsMedia;
  var isSortableComputed = sortable === true || sortable === false ? sortable : listIsSortable;
  var isSortableOppositeComputed = isSortableComputed && (sortableOpposite || listIsSortableOpposite);

  if (!listIsSimple) {
    // Item Content
    itemContentEl = /*#__PURE__*/_react.default.createElement(_listItemContent.default, _extends({}, props, {
      slots: slots,
      onChange: onChange,
      onClick: link || href || accordionItem || smartSelect ? undefined : onClick,
      isMediaComputed: isMediaComputed,
      isSortableComputed: isSortableComputed,
      isSortableOppositeComputed: isSortableOppositeComputed
    })); // Link

    if (link || href || accordionItem || smartSelect) {
      var linkAttrs = _extends({
        href: link === true ? '' : link || href,
        target: target,
        'data-tab': (0, _utils.isStringProp)(tabLink) && tabLink || undefined
      }, (0, _mixins.routerAttrs)(props), (0, _mixins.actionsAttrs)(props));

      var linkClasses = (0, _utils.classNames)({
        'item-link': true,
        'smart-select': smartSelect,
        'tab-link': tabLink || tabLink === '',
        'tab-link-active': tabLinkActive,
        'item-selected': selected
      }, (0, _mixins.routerClasses)(props), (0, _mixins.actionsClasses)(props));
      linkEl = /*#__PURE__*/_react.default.createElement("a", _extends({
        ref: linkElRef,
        className: linkClasses
      }, linkAttrs, {
        onClick: onClick
      }), itemContentEl);
    }
  }

  var liClasses = (0, _utils.classNames)(className, {
    'item-divider': divider,
    'list-group-title': groupTitle,
    'media-item': isMediaComputed,
    swipeout: swipeout,
    'accordion-item': accordionItem,
    'accordion-item-opened': accordionItemOpened,
    disabled: disabled && !(radio || checkbox),
    'no-chevron': noChevron,
    'chevron-center': chevronCenter,
    'disallow-sorting': sortable === false
  }, (0, _mixins.colorClasses)(props));

  if (divider || groupTitle) {
    return /*#__PURE__*/_react.default.createElement("li", {
      ref: elRef,
      id: id,
      style: style,
      className: liClasses,
      "data-virtual-list-index": virtualListIndex,
      onClick: onClick
    }, /*#__PURE__*/_react.default.createElement("span", null, title, children));
  }

  if (listIsSimple) {
    return /*#__PURE__*/_react.default.createElement("li", {
      ref: elRef,
      id: id,
      style: style,
      className: liClasses,
      "data-virtual-list-index": virtualListIndex,
      onClick: onClick
    }, title, children);
  }

  var linkItemEl = link || href || smartSelect || accordionItem ? linkEl : itemContentEl;
  return /*#__PURE__*/_react.default.createElement("li", _extends({
    ref: elRef,
    id: id,
    style: style,
    className: liClasses,
    "data-virtual-list-index": virtualListIndex
  }, extraAttrs), slots['root-start'], swipeout ? /*#__PURE__*/_react.default.createElement("div", {
    className: "swipeout-content"
  }, linkItemEl) : linkItemEl, isSortableComputed && sortable !== false && !isSortableOppositeComputed && /*#__PURE__*/_react.default.createElement("div", {
    className: "sortable-handler"
  }), (swipeout || accordionItem) && slots.default, slots.root, slots['root-end']);
});
ListItem.displayName = 'f7-list-item';
var _default = ListItem;
exports.default = _default;
