// framework7-react/cjs/components/messages.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _f = require("../shared/f7");

var _watchProp = require("../shared/watch-prop");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }


var Messages = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _f7Messages = (0, _react.useRef)(null);

  var mounted = (0, _react.useRef)(false);
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      _props$autoLayout = props.autoLayout,
      autoLayout = _props$autoLayout === void 0 ? false : _props$autoLayout,
      _props$messages = props.messages,
      messages = _props$messages === void 0 ? [] : _props$messages,
      _props$newMessagesFir = props.newMessagesFirst,
      newMessagesFirst = _props$newMessagesFir === void 0 ? false : _props$newMessagesFir,
      _props$scrollMessages = props.scrollMessages,
      scrollMessages = _props$scrollMessages === void 0 ? true : _props$scrollMessages,
      _props$scrollMessages2 = props.scrollMessagesOnEdge,
      scrollMessagesOnEdge = _props$scrollMessages2 === void 0 ? true : _props$scrollMessages2,
      firstMessageRule = props.firstMessageRule,
      lastMessageRule = props.lastMessageRule,
      tailMessageRule = props.tailMessageRule,
      sameNameMessageRule = props.sameNameMessageRule,
      sameHeaderMessageRule = props.sameHeaderMessageRule,
      sameFooterMessageRule = props.sameFooterMessageRule,
      sameAvatarMessageRule = props.sameAvatarMessageRule,
      customClassMessageRule = props.customClassMessageRule,
      renderMessage = props.renderMessage,
      _props$typing = props.typing,
      typing = _props$typing === void 0 ? false : _props$typing,
      _props$init = props.init,
      init = _props$init === void 0 ? true : _props$init;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  var childrenBeforeUpdated = (0, _react.useRef)(null);
  var reactChildrenBefore = (0, _react.useRef)(children ? _react.default.Children.count(children) : 0);
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7Messages: function f7Messages() {
        return _f7Messages.current;
      }
    };
  });

  var onMount = function onMount() {
    if (!init) return;
    (0, _f.f7ready)(function () {
      _f7Messages.current = _f.f7.messages.create((0, _utils.noUndefinedProps)({
        el: elRef.current,
        autoLayout: autoLayout,
        messages: messages,
        newMessagesFirst: newMessagesFirst,
        scrollMessages: scrollMessages,
        scrollMessagesOnEdge: scrollMessagesOnEdge,
        firstMessageRule: firstMessageRule,
        lastMessageRule: lastMessageRule,
        tailMessageRule: tailMessageRule,
        sameNameMessageRule: sameNameMessageRule,
        sameHeaderMessageRule: sameHeaderMessageRule,
        sameFooterMessageRule: sameFooterMessageRule,
        sameAvatarMessageRule: sameAvatarMessageRule,
        customClassMessageRule: customClassMessageRule,
        renderMessage: renderMessage
      }));

      if (typing) {
        _f7Messages.current.showTyping();
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7Messages.current && _f7Messages.current.destroy) _f7Messages.current.destroy();
    _f7Messages.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var currentChildrenLength = children ? _react.default.Children.count(children) : 0;

  if (_f7Messages.current && scrollMessages) {
    var beforeChildrenLength = reactChildrenBefore.current || 0;

    if (currentChildrenLength !== beforeChildrenLength) {
      _f7Messages.current.setScrollData();
    }
  }

  reactChildrenBefore.current = currentChildrenLength;
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    var wasMounted = mounted.current;
    mounted.current = true;
    if (!init || !elRef.current) return;
    var childElements = elRef.current.children;
    if (!childElements) return;
    var childrenAfterUpdated = childElements.length;

    if (!wasMounted) {
      for (var i = 0; i < childElements.length; i += 1) {
        childElements[i].classList.add('message-appeared');
      }

      return;
    }

    for (var _i = 0; _i < childElements.length; _i += 1) {
      if (!childElements[_i].classList.contains('message-appeared')) {
        childElements[_i].classList.add('message-appear-from-bottom');
      }
    }

    if (_f7Messages.current) {
      if (_f7Messages.current.layout && autoLayout) {
        _f7Messages.current.layout();
      }

      if (childrenBeforeUpdated.current !== childrenAfterUpdated && _f7Messages.current.scroll && _f7Messages.current.scrollData && scrollMessages) {
        _f7Messages.current.scrollWithEdgeCheck(true);
      }
    }

    childrenBeforeUpdated.current = childrenAfterUpdated;
  });
  (0, _watchProp.watchProp)(typing, function (newValue) {
    if (!_f7Messages.current) return;
    if (newValue) _f7Messages.current.showTyping();else _f7Messages.current.hideTyping();
  });
  var classes = (0, _utils.classNames)(className, 'messages', (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
Messages.displayName = 'f7-messages';
var _default = Messages;
exports.default = _default;
