// framework7-react/cjs/components/message.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var Message = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      text = props.text,
      name = props.name,
      avatar = props.avatar,
      _props$type = props.type,
      type = _props$type === void 0 ? 'sent' : _props$type,
      image = props.image,
      header = props.header,
      footer = props.footer,
      textHeader = props.textHeader,
      textFooter = props.textFooter,
      first = props.first,
      last = props.last,
      tail = props.tail,
      sameName = props.sameName,
      sameHeader = props.sameHeader,
      sameFooter = props.sameFooter,
      sameAvatar = props.sameAvatar,
      typing = props.typing;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);

  var onClick = function onClick(event) {
    (0, _utils.emit)(props, 'click', event);
  };

  var onNameClick = function onNameClick(event) {
    (0, _utils.emit)(props, 'clickName', event);
  };

  var onTextClick = function onTextClick(event) {
    (0, _utils.emit)(props, 'clickText', event);
  };

  var onAvatarClick = function onAvatarClick(event) {
    (0, _utils.emit)(props, 'clickAvatar', event);
  };

  var onHeaderClick = function onHeaderClick(event) {
    (0, _utils.emit)(props, 'clickHeader', event);
  };

  var onFooterClick = function onFooterClick(event) {
    (0, _utils.emit)(props, 'clickFooter', event);
  };

  var onBubbleClick = function onBubbleClick(event) {
    (0, _utils.emit)(props, 'clickBubble', event);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });
  var slots = (0, _utils.getSlots)(props);
  var classes = (0, _utils.classNames)(className, 'message', {
    'message-sent': type === 'sent',
    'message-received': type === 'received',
    'message-typing': typing,
    'message-first': first,
    'message-last': last,
    'message-tail': tail,
    'message-same-name': sameName,
    'message-same-header': sameHeader,
    'message-same-footer': sameFooter,
    'message-same-avatar': sameAvatar
  }, (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs, {
    onClick: onClick
  }), slots.start, (avatar || slots.avatar) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-avatar",
    style: {
      backgroundImage: avatar && "url(" + avatar + ")"
    },
    onClick: onAvatarClick
  }, slots.avatar), /*#__PURE__*/_react.default.createElement("div", {
    className: "message-content"
  }, slots['content-start'], (slots.name || name) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-name",
    onClick: onNameClick
  }, name, slots.name), (slots.header || header) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-header",
    onClick: onHeaderClick
  }, header, slots.header), /*#__PURE__*/_react.default.createElement("div", {
    className: "message-bubble",
    onClick: onBubbleClick
  }, slots['bubble-start'], (slots.image || image) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-image"
  }, slots.image || /*#__PURE__*/_react.default.createElement("img", {
    src: image
  })), (slots['text-header'] || textHeader) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-text-header"
  }, textHeader, slots['text-header']), (slots.text || text || typing) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-text",
    onClick: onTextClick
  }, text, slots.text, typing && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-typing-indicator"
  }, /*#__PURE__*/_react.default.createElement("div", null), /*#__PURE__*/_react.default.createElement("div", null), /*#__PURE__*/_react.default.createElement("div", null))), (slots['text-footer'] || textFooter) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-text-footer"
  }, textFooter, slots['text-footer']), slots['bubble-end'], slots.default), (slots.footer || footer) && /*#__PURE__*/_react.default.createElement("div", {
    className: "message-footer",
    onClick: onFooterClick
  }, footer, slots.footer), slots['content-end']), slots.end);
});
Message.displayName = 'f7-message';
var _default = Message;
exports.default = _default;
