// framework7-react/cjs/components/nav-left.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _useTheme = require("../shared/use-theme");

var _link = _interopRequireDefault(require("./link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var NavLeft = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      backLink = props.backLink,
      backLinkUrl = props.backLinkUrl,
      backLinkForce = props.backLinkForce,
      backLinkShowText = props.backLinkShowText,
      sliding = props.sliding;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);

  var onBackClick = function onBackClick(event) {
    (0, _utils.emit)(props, 'backClick clickBack', event);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });
  var theme = (0, _useTheme.useTheme)();
  var linkEl;
  var needBackLinkText = backLinkShowText;
  if (typeof needBackLinkText === 'undefined') needBackLinkText = !theme.md;

  if (backLink) {
    var text = backLink !== true && needBackLinkText ? backLink : undefined;
    linkEl = /*#__PURE__*/_react.default.createElement(_link.default, {
      href: backLinkUrl || '#',
      back: true,
      icon: "icon-back",
      force: backLinkForce || undefined,
      className: !text ? 'icon-only' : undefined,
      text: text,
      onClick: onBackClick
    });
  }

  var classes = (0, _utils.classNames)(className, 'left', {
    sliding: sliding
  }, (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), linkEl, children);
});
NavLeft.displayName = 'f7-nav-left';
var _default = NavLeft;
exports.default = _default;
