// framework7-react/cjs/components/page-content.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _preloader = _interopRequireDefault(require("./preloader"));

var _useTab = require("../shared/use-tab");

var _f = require("../shared/f7");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var PageContent = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      tab = props.tab,
      tabActive = props.tabActive,
      ptr = props.ptr,
      ptrDistance = props.ptrDistance,
      _props$ptrPreloader = props.ptrPreloader,
      ptrPreloader = _props$ptrPreloader === void 0 ? true : _props$ptrPreloader,
      ptrBottom = props.ptrBottom,
      ptrMousewheel = props.ptrMousewheel,
      infinite = props.infinite,
      infiniteTop = props.infiniteTop,
      infiniteDistance = props.infiniteDistance,
      _props$infinitePreloa = props.infinitePreloader,
      infinitePreloader = _props$infinitePreloa === void 0 ? true : _props$infinitePreloa,
      hideBarsOnScroll = props.hideBarsOnScroll,
      hideNavbarOnScroll = props.hideNavbarOnScroll,
      hideToolbarOnScroll = props.hideToolbarOnScroll,
      messagesContent = props.messagesContent,
      loginScreen = props.loginScreen;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);

  var onPtrPullStart = function onPtrPullStart(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'ptrPullStart');
  };

  var onPtrPullMove = function onPtrPullMove(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'ptrPullMove');
  };

  var onPtrPullEnd = function onPtrPullEnd(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'ptrPullEnd');
  };

  var onPtrRefresh = function onPtrRefresh(el, done) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'ptrRefresh', done);
  };

  var onPtrDone = function onPtrDone(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'ptrDone');
  };

  var onInfinite = function onInfinite(el) {
    if (elRef.current !== el) return;
    (0, _utils.emit)(props, 'infinite');
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });
  (0, _useTab.useTab)(elRef, props);

  var attachEvents = function attachEvents() {
    (0, _f.f7ready)(function () {
      if (ptr) {
        _f.f7.on('ptrPullStart', onPtrPullStart);

        _f.f7.on('ptrPullMove', onPtrPullMove);

        _f.f7.on('ptrPullEnd', onPtrPullEnd);

        _f.f7.on('ptrRefresh', onPtrRefresh);

        _f.f7.on('ptrDone', onPtrDone);
      }

      if (infinite) {
        _f.f7.on('infinite', onInfinite);
      }
    });
  };

  var detachEvents = function detachEvents() {
    if (!_f.f7) return;

    _f.f7.off('ptrPullStart', onPtrPullStart);

    _f.f7.off('ptrPullMove', onPtrPullMove);

    _f.f7.off('ptrPullEnd', onPtrPullEnd);

    _f.f7.off('ptrRefresh', onPtrRefresh);

    _f.f7.off('ptrDone', onPtrDone);

    _f.f7.off('infinite', onInfinite);
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    attachEvents();
    return detachEvents;
  });
  var ptrEl;
  var infiniteEl;

  if (ptr && ptrPreloader) {
    ptrEl = /*#__PURE__*/_react.default.createElement("div", {
      className: "ptr-preloader"
    }, /*#__PURE__*/_react.default.createElement(_preloader.default, null), /*#__PURE__*/_react.default.createElement("div", {
      className: "ptr-arrow"
    }));
  }

  if (infinite && infinitePreloader) {
    infiniteEl = /*#__PURE__*/_react.default.createElement(_preloader.default, {
      className: "infinite-scroll-preloader"
    });
  }

  var classes = (0, _utils.classNames)(className, 'page-content', {
    tab: tab,
    'tab-active': tabActive,
    'ptr-content': ptr,
    'ptr-bottom': ptrBottom,
    'infinite-scroll-content': infinite,
    'infinite-scroll-top': infiniteTop,
    'hide-bars-on-scroll': hideBarsOnScroll,
    'hide-navbar-on-scroll': hideNavbarOnScroll,
    'hide-toolbar-on-scroll': hideToolbarOnScroll,
    'messages-content': messagesContent,
    'login-screen-content': loginScreen
  }, (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    "data-ptr-distance": ptrDistance || undefined,
    "data-ptr-mousewheel": ptrMousewheel || undefined,
    "data-infinite-distance": infiniteDistance || undefined,
    ref: elRef
  }, extraAttrs), ptrBottom ? null : ptrEl, infiniteTop ? infiniteEl : null, children, infiniteTop ? null : infiniteEl, ptrBottom ? ptrEl : null);
});
PageContent.displayName = 'f7-page-content';
var _default = PageContent;
exports.default = _default;
