// framework7-react/cjs/components/panel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _f = require("../shared/f7");

var _watchProp = require("../shared/watch-prop");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }


var Panel = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _classNames;

  var _f7Panel = (0, _react.useRef)(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      side = props.side,
      effect = props.effect,
      cover = props.cover,
      reveal = props.reveal,
      push = props.push,
      left = props.left,
      opened = props.opened,
      resizable = props.resizable,
      _props$backdrop = props.backdrop,
      backdrop = _props$backdrop === void 0 ? true : _props$backdrop,
      backdropEl = props.backdropEl,
      containerEl = props.containerEl,
      visibleBreakpoint = props.visibleBreakpoint,
      collapsedBreakpoint = props.collapsedBreakpoint,
      swipe = props.swipe,
      swipeNoFollow = props.swipeNoFollow,
      swipeOnlyClose = props.swipeOnlyClose,
      _props$swipeActiveAre = props.swipeActiveArea,
      swipeActiveArea = _props$swipeActiveAre === void 0 ? 0 : _props$swipeActiveAre,
      _props$swipeThreshold = props.swipeThreshold,
      swipeThreshold = _props$swipeThreshold === void 0 ? 0 : _props$swipeThreshold;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  var isOpened = (0, _react.useRef)(false);
  var isClosing = (0, _react.useRef)(false);
  var isCollapsed = (0, _react.useRef)(false);
  var isBreakpoint = (0, _react.useRef)(false);

  var onOpen = function onOpen(event) {
    isOpened.current = true;
    isClosing.current = false;
    (0, _utils.emit)(props, 'panelOpen', event);
  };

  var onOpened = function onOpened(event) {
    (0, _utils.emit)(props, 'panelOpened', event);
  };

  var onClose = function onClose(event) {
    isOpened.current = false;
    isClosing.current = true;
    (0, _utils.emit)(props, 'panelClose', event);
  };

  var onClosed = function onClosed(event) {
    isClosing.current = false;
    (0, _utils.emit)(props, 'panelClosed', event);
  };

  var onBackdropClick = function onBackdropClick(event) {
    (0, _utils.emit)(props, 'click panelBackdropClick', event);
  };

  var onSwipe = function onSwipe(event) {
    (0, _utils.emit)(props, 'panelSwipe', event);
  };

  var onSwipeOpen = function onSwipeOpen(event) {
    (0, _utils.emit)(props, 'panelSwipeOpen', event);
  };

  var onBreakpoint = function onBreakpoint(event) {
    isBreakpoint.current = true;
    isCollapsed.current = false;
    (0, _utils.emit)(props, 'panelBreakpoint', event);
  };

  var onCollapsedBreakpoint = function onCollapsedBreakpoint(event) {
    isBreakpoint.current = false;
    isCollapsed.current = true;
    (0, _utils.emit)(props, 'panelCollapsedBreakpoint', event);
  };

  var onResize = function onResize() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _utils.emit.apply(void 0, [props, 'panelResize'].concat(args));
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7Panel: function f7Panel() {
        return _f7Panel.current;
      }
    };
  });
  (0, _watchProp.watchProp)(resizable, function (newValue) {
    if (!_f7Panel.current) return;
    if (newValue) _f7Panel.current.enableResizable();else _f7Panel.current.disableResizable();
  });
  (0, _watchProp.watchProp)(opened, function (newValue) {
    if (!_f7Panel.current) return;

    if (newValue) {
      _f7Panel.current.open();
    } else {
      _f7Panel.current.close();
    }
  });

  var modalEvents = function modalEvents(method) {
    if (!_f7Panel.current) return;

    _f7Panel.current[method]('open', onOpen);

    _f7Panel.current[method]('opened', onOpened);

    _f7Panel.current[method]('close', onClose);

    _f7Panel.current[method]('closed', onClosed);

    _f7Panel.current[method]('backdropClick', onBackdropClick);

    _f7Panel.current[method]('swipe', onSwipe);

    _f7Panel.current[method]('swipeOpen', onSwipeOpen);

    _f7Panel.current[method]('collapsedBreakpoint', onCollapsedBreakpoint);

    _f7Panel.current[method]('breakpoint', onBreakpoint);

    _f7Panel.current[method]('resize', onResize);
  };

  var onMount = function onMount() {
    (0, _f.f7ready)(function () {
      var $ = _f.f7.$;
      if (!$) return;

      if ($('.panel-backdrop').length === 0) {
        $('<div class="panel-backdrop"></div>').insertBefore(elRef.current);
      }

      var params = (0, _utils.noUndefinedProps)({
        el: elRef.current,
        resizable: resizable,
        backdrop: backdrop,
        backdropEl: backdropEl,
        containerEl: containerEl,
        visibleBreakpoint: visibleBreakpoint,
        collapsedBreakpoint: collapsedBreakpoint,
        swipe: swipe,
        swipeNoFollow: swipeNoFollow,
        swipeOnlyClose: swipeOnlyClose,
        swipeActiveArea: swipeActiveArea,
        swipeThreshold: swipeThreshold
      });
      _f7Panel.current = _f.f7.panel.create(params);
      modalEvents('on');

      if (opened) {
        _f7Panel.current.open(false);
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7Panel.current && _f7Panel.current.destroy) {
      _f7Panel.current.destroy();
    }

    _f7Panel.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    modalEvents('on');
    return function () {
      modalEvents('off');
    };
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var sideComputed = side || (left ? 'left' : 'right'); // eslint-disable-next-line

  var effectComputed = effect || (reveal ? 'reveal' : push ? 'push' : 'cover');
  var classes = (0, _utils.classNames)(className, 'panel', (_classNames = {
    'panel-in': isOpened.current && !isClosing.current && !isBreakpoint.current,
    'panel-in-breakpoint': isBreakpoint.current,
    'panel-in-collapsed': isCollapsed.current,
    'panel-resizable': resizable
  }, _classNames["panel-" + sideComputed] = sideComputed, _classNames["panel-" + effectComputed] = effectComputed, _classNames), (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children, resizable && /*#__PURE__*/_react.default.createElement("div", {
    className: "panel-resize-handler"
  }));
});
Panel.displayName = 'f7-panel';
var _default = Panel;
exports.default = _default;
