// framework7-react/cjs/components/popup.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _f = require("../shared/f7");

var _mixins = require("../shared/mixins");

var _modalStateClasses = require("../shared/modal-state-classes");

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _watchProp = require("../shared/watch-prop");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }


var Popup = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _f7Popup = (0, _react.useRef)(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      tabletFullscreen = props.tabletFullscreen,
      push = props.push,
      opened = props.opened,
      closeByBackdropClick = props.closeByBackdropClick,
      backdrop = props.backdrop,
      backdropEl = props.backdropEl,
      animate = props.animate,
      closeOnEscape = props.closeOnEscape,
      _props$swipeToClose = props.swipeToClose,
      swipeToClose = _props$swipeToClose === void 0 ? false : _props$swipeToClose,
      swipeHandler = props.swipeHandler,
      containerEl = props.containerEl;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  var isOpened = (0, _react.useRef)(opened);
  var isClosing = (0, _react.useRef)(false);

  var onSwipeStart = function onSwipeStart(instance) {
    (0, _utils.emit)(props, 'popupSwipeStart', instance);
  };

  var onSwipeMove = function onSwipeMove(instance) {
    (0, _utils.emit)(props, 'popupSwipeMove', instance);
  };

  var onSwipeEnd = function onSwipeEnd(instance) {
    (0, _utils.emit)(props, 'popupSwipeEnd', instance);
  };

  var onSwipeClose = function onSwipeClose(instance) {
    (0, _utils.emit)(props, 'popupSwipeClose', instance);
  };

  var onOpen = function onOpen(instance) {
    isOpened.current = true;
    isClosing.current = false;
    (0, _utils.emit)(props, 'popupOpen', instance);
  };

  var onOpened = function onOpened(instance) {
    (0, _utils.emit)(props, 'popupOpened', instance);
  };

  var onClose = function onClose(instance) {
    isOpened.current = false;
    isClosing.current = true;
    (0, _utils.emit)(props, 'popupClose', instance);
  };

  var onClosed = function onClosed(instance) {
    isClosing.current = false;
    (0, _utils.emit)(props, 'popupClosed', instance);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7Popup: function f7Popup() {
        return _f7Popup.current;
      }
    };
  });
  (0, _watchProp.watchProp)(opened, function (value) {
    if (!_f7Popup.current) return;

    if (value) {
      _f7Popup.current.open();
    } else {
      _f7Popup.current.close();
    }
  });

  var modalEvents = function modalEvents(method) {
    if (!_f7Popup.current) return;

    _f7Popup.current[method]('swipeStart', onSwipeStart);

    _f7Popup.current[method]('swipeMove', onSwipeMove);

    _f7Popup.current[method]('swipeEnd', onSwipeEnd);

    _f7Popup.current[method]('swipeClose', onSwipeClose);

    _f7Popup.current[method]('open', onOpen);

    _f7Popup.current[method]('opened', onOpened);

    _f7Popup.current[method]('close', onClose);

    _f7Popup.current[method]('closed', onClosed);
  };

  var onMount = function onMount() {
    if (!elRef.current) return;
    var popupParams = {
      el: elRef.current
    };
    if ('closeByBackdropClick' in props) popupParams.closeByBackdropClick = closeByBackdropClick;
    if ('closeOnEscape' in props) popupParams.closeOnEscape = closeOnEscape;
    if ('animate' in props) popupParams.animate = animate;
    if ('backdrop' in props) popupParams.backdrop = backdrop;
    if ('backdropEl' in props) popupParams.backdropEl = backdropEl;
    if ('swipeToClose' in props) popupParams.swipeToClose = swipeToClose;
    if ('swipeHandler' in props) popupParams.swipeHandler = swipeHandler;
    if ('containerEl' in props) popupParams.containerEl = containerEl;
    (0, _f.f7ready)(function () {
      _f7Popup.current = _f.f7.popup.create(popupParams);
      modalEvents('on');

      if (opened) {
        _f7Popup.current.open(false);
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7Popup.current) {
      _f7Popup.current.destroy();
    }

    _f7Popup.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    modalEvents('on');
    return function () {
      modalEvents('off');
    };
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var classes = (0, _utils.classNames)(className, 'popup', {
    'popup-tablet-fullscreen': tabletFullscreen,
    'popup-push': push
  }, (0, _modalStateClasses.modalStateClasses)({
    isOpened: isOpened,
    isClosing: isClosing
  }), (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
Popup.displayName = 'f7-popup';
var _default = Popup;
exports.default = _default;
