// framework7-react/cjs/components/range.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _f = require("../shared/f7");

var _watchProp = require("../shared/watch-prop");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }


var Range = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _f7Range = (0, _react.useRef)(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      _props$init = props.init,
      init = _props$init === void 0 ? true : _props$init,
      _props$value = props.value,
      value = _props$value === void 0 ? 0 : _props$value,
      _props$min = props.min,
      min = _props$min === void 0 ? 0 : _props$min,
      _props$max = props.max,
      max = _props$max === void 0 ? 100 : _props$max,
      _props$step = props.step,
      step = _props$step === void 0 ? 1 : _props$step,
      _props$label = props.label,
      label = _props$label === void 0 ? false : _props$label,
      _props$dual = props.dual,
      dual = _props$dual === void 0 ? false : _props$dual,
      _props$vertical = props.vertical,
      vertical = _props$vertical === void 0 ? false : _props$vertical,
      _props$verticalRevers = props.verticalReversed,
      verticalReversed = _props$verticalRevers === void 0 ? false : _props$verticalRevers,
      _props$draggableBar = props.draggableBar,
      draggableBar = _props$draggableBar === void 0 ? true : _props$draggableBar,
      formatLabel = props.formatLabel,
      _props$scale = props.scale,
      scale = _props$scale === void 0 ? false : _props$scale,
      _props$scaleSteps = props.scaleSteps,
      scaleSteps = _props$scaleSteps === void 0 ? 5 : _props$scaleSteps,
      _props$scaleSubSteps = props.scaleSubSteps,
      scaleSubSteps = _props$scaleSubSteps === void 0 ? 0 : _props$scaleSubSteps,
      formatScaleLabel = props.formatScaleLabel,
      _props$limitKnobPosit = props.limitKnobPosition,
      limitKnobPosition = _props$limitKnobPosit === void 0 ? undefined : _props$limitKnobPosit,
      name = props.name,
      input = props.input,
      inputId = props.inputId,
      disabled = props.disabled;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7Range: function f7Range() {
        return _f7Range.current;
      }
    };
  });
  (0, _watchProp.watchProp)(value, function (newValue) {
    if (!_f7Range.current) return;
    var rangeValue = _f7Range.current.value;

    if (Array.isArray(newValue) && Array.isArray(rangeValue)) {
      if (rangeValue[0] !== newValue[0] || rangeValue[1] !== newValue[1]) {
        _f7Range.current.setValue(newValue);
      }
    } else {
      _f7Range.current.setValue(newValue);
    }
  });

  var onChange = function onChange(range, val) {
    (0, _utils.emit)(props, 'rangeChange', val);
  };

  var onChanged = function onChanged(range, val) {
    (0, _utils.emit)(props, 'rangeChanged', val);
  };

  var rangeEvents = function rangeEvents(method) {
    if (!_f7Range.current) return;

    _f7Range.current[method]('change', onChange);

    _f7Range.current[method]('changed', onChanged);
  };

  var onMount = function onMount() {
    (0, _f.f7ready)(function () {
      if (!init || !elRef.current) return;
      _f7Range.current = _f.f7.range.create((0, _utils.noUndefinedProps)({
        el: elRef.current,
        value: value,
        min: min,
        max: max,
        step: step,
        label: label,
        dual: dual,
        draggableBar: draggableBar,
        vertical: vertical,
        verticalReversed: verticalReversed,
        formatLabel: formatLabel,
        scale: scale,
        scaleSteps: scaleSteps,
        scaleSubSteps: scaleSubSteps,
        formatScaleLabel: formatScaleLabel,
        limitKnobPosition: limitKnobPosition
      }));
      rangeEvents('on');
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7Range.current && _f7Range.current.destroy) _f7Range.current.destroy();
    _f7Range.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    rangeEvents('on');
    return function () {
      rangeEvents('off');
    };
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var classes = (0, _utils.classNames)(className, 'range-slider', {
    'range-slider-horizontal': !vertical,
    'range-slider-vertical': vertical,
    'range-slider-vertical-reversed': vertical && verticalReversed,
    disabled: disabled
  }, (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    ref: elRef,
    id: id,
    style: style,
    className: classes
  }, extraAttrs), input && /*#__PURE__*/_react.default.createElement("input", {
    type: "range",
    name: name,
    id: inputId
  }), children);
});
Range.displayName = 'f7-range';
var _default = Range;
exports.default = _default;
