// framework7-react/cjs/components/toggle.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _f = require("../shared/f7");

var _watchProp = require("../shared/watch-prop");

var _useTooltip = require("../shared/use-tooltip");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }


var Toggle = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _f7Toggle = (0, _react.useRef)(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      _props$init = props.init,
      init = _props$init === void 0 ? true : _props$init,
      checked = props.checked,
      defaultChecked = props.defaultChecked,
      disabled = props.disabled,
      readonly = props.readonly,
      name = props.name,
      value = props.value;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  var inputElRef = (0, _react.useRef)(null);

  var onChange = function onChange(event) {
    (0, _utils.emit)(props, 'change', event);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      f7Toggle: function f7Toggle() {
        return _f7Toggle.current;
      }
    };
  });
  (0, _useTooltip.useTooltip)(elRef, props);
  (0, _watchProp.watchProp)(checked, function (newValue) {
    if (!_f7Toggle.current) return;
    _f7Toggle.current.checked = newValue;
  });

  var onToggleChange = function onToggleChange(toggleInstance) {
    (0, _utils.emit)(props, 'toggleChange', toggleInstance.checked);
  };

  var toggleEvents = function toggleEvents(method) {
    if (!_f7Toggle.current) return;

    _f7Toggle.current[method]('toggleChange', onToggleChange);
  };

  var onMount = function onMount() {
    (0, _f.f7ready)(function () {
      if (!init || !elRef.current) return;
      _f7Toggle.current = _f.f7.toggle.create({
        el: elRef.current
      });
      toggleEvents('on');
    });
  };

  var onDestroy = function onDestroy() {
    if (_f7Toggle.current && _f7Toggle.current.destroy && _f7Toggle.current.$el) {
      _f7Toggle.current.destroy();
    }

    _f7Toggle.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    toggleEvents('on');

    if (inputElRef.current) {
      inputElRef.current.addEventListener('change', onChange);
    }

    return function () {
      toggleEvents('off');

      if (inputElRef.current) {
        inputElRef.current.removeEventListener('change', onChange);
      }
    };
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var labelClasses = (0, _utils.classNames)('toggle', className, {
    disabled: disabled
  }, (0, _mixins.colorClasses)(props));

  var inputEl = /*#__PURE__*/_react.default.createElement("input", {
    ref: inputElRef,
    type: "checkbox",
    name: name,
    disabled: disabled,
    readOnly: readonly,
    checked: checked,
    defaultChecked: defaultChecked,
    value: value,
    onChange: function onChange() {}
  });

  return /*#__PURE__*/_react.default.createElement("label", _extends({
    id: id,
    style: style,
    className: labelClasses,
    ref: elRef
  }, extraAttrs), inputEl, /*#__PURE__*/_react.default.createElement("span", {
    className: "toggle-icon"
  }));
});
Toggle.displayName = 'f7-toggle';
var _default = Toggle;
exports.default = _default;
